backend/fofn.go

```diff
@@ -182,16 +193,31 @@ func (s *Server) createRulesToAdd(user string, rule db.Rule, files []string) ([]
 	return rulesToAdd, nil
 }
 
-func (s *Server) claimAndCreateDirRules(file, user string) (*ruletree.DirRules, error) {
+func (s *Server) claimAndCreateDirRules(file, user string, dirDetails dirDetails,
+	detailsSet map[*ruletree.DirRules]struct{}) (*ruletree.DirRules, error) {
 	fileDir, _ := splitDir(file)
 	dirRules := s.directoryRules[fileDir]
 
 	if dirRules == nil {
-		if err := s.claimDirectory(fileDir, user); err != nil {
+		if err := s.claimDirectory(fileDir, user, dirDetails); err != nil {
 			return nil, err
 		}
 
 		dirRules = s.directoryRules[fileDir]
+
+		return dirRules, nil
+	}
+
+	if _, ok := detailsSet[dirRules]; !ok {
```

Comment:
It's nicer if this is a map of bool, then you can just say `if detailsSet[dirRules]`